### PR TITLE
CDAP-8032 retry when connect fails after service discovery

### PR DESCRIPTION
**Pull request by albertshau**
Refactoring several clients that perform service discovery
then http requests so that they use a single common class. This
client will throw ServiceUnavailableException if the service
cannot be discovered, if discovery succeeded but a ConnectException
was thrown while making the request, or if the response itself
was a 503 service unavailable.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-RUT555

File: cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteClient.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * Discovers a remote service and resolves URLs to that service.
+ */
+public class RemoteClient {
+  private final Supplier<EndpointStrategy> endpointStrategySupplier;
+  private final HttpRequestConfig httpRequestConfig;
+  private final String discoverableServiceName;
+  private final String basePath;
+
+  public RemoteClient(final DiscoveryServiceClient discoveryClient,
+                      final String discoverableServiceName,
```
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteClient.java]: nit: no need to have this on a separate line.

File: cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteClient.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        return new RandomEndpointStrategy(discoveryClient.discover(discoverableServiceName));
+      }
+    });
+    String cleanBasePath = basePath.startsWith("/") ? basePath : "/" + basePath;
+    this.basePath = cleanBasePath.endsWith("/") ? cleanBasePath : cleanBasePath + "/";
+  }
+
+  /**
+   * Discover the service address and use it and the base path and specified resource to generate the URL
+   * to use when making a request to the service.
```
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteClient.java]: Description looks confusing for the method `requestBuilder`

File: cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteOpsClient.java [hunk 3]
```diff
@@ -63,43 +48,39 @@
     .registerTypeAdapter(WorkflowTokenNodeDetail.class, new WorkflowTokenNodeDetailCodec())
     .create();
 
-  private final Supplier<EndpointStrategy> endpointStrategySupplier;
-  private final HttpRequestConfig httpRequestConfig;
-  private final String discoverableServiceName;
+  private final RemoteClient remoteClient;
 
   protected RemoteOpsClient(CConfiguration cConf, final DiscoveryServiceClient discoveryClient,
```
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/internal/remote/RemoteOpsClient.java]: not this PR introduced it, but seems like cConf in unused

File: cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java [hunk 4]
```diff
@@ -16,71 +16,44 @@
...
  * Implementation of {@link NamespaceQueryAdmin} that pings the internal endpoints in a {@link HttpHandler} in remote
  * system service.
  */
 public class RemoteNamespaceQueryClient extends AbstractNamespaceQueryClient {
-  private final Supplier<EndpointStrategy> endpointStrategySupplier;
+  private final RemoteClient remoteClient;
   private final AuthenticationContext authenticationContext;
 
   @Inject
   RemoteNamespaceQueryClient(final DiscoveryServiceClient discoveryClient, CConfiguration cConf,
```
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:42]: seems like this cConf is also not used.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 5]
```diff
@@ -301,61 +272,40 @@ void deleteModules() throws DatasetManagementException {
...
 
   private HttpResponse doPut(String resource, String body) throws DatasetManagementException {
-    return doRequest(HttpMethod.PUT, resource, null, body);
+    HttpRequest request = addUserIdHeader(remoteClient.requestBuilder(HttpMethod.PUT, resource)
+      .withBody(body))
+      .build();
+    try {
+      return remoteClient.execute(request);
+    } catch (IOException e) {
+      throw new DatasetManagementException(remoteClient.createErrorMessage(request, body));
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: better pass `e` to DatasetManagementException.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 6]
```diff
@@ -301,61 +272,40 @@ void deleteModules() throws DatasetManagementException {
...
+      return remoteClient.execute(request);
+    } catch (ConnectException e) {
+      throw new ServiceUnavailableException(Constants.Service.DATASET_MANAGER, e);
     } catch (IOException e) {
-      throw new DatasetManagementException(
-        String.format("Error during talking to Dataset Service at %s while doing %s with headers %s and body %s",
-                      url, method,
-                      headers == null ? "null" : Joiner.on(",").withKeyValueSeparator("=").join(headers.entries()),
-                      body == null ? "null" : body), e);
+      throw new DatasetManagementException(remoteClient.createErrorMessage(request, null));
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: better pass `e` to DatasetManagementException

**Review by sagarkapare — COMMENTED** [review]
Mostly looks good. few comments.

**Review by sagarkapare — APPROVED** [review]
LGTM

**Pull request merged**